File: least_squares.cpp
//program for calculation
//using least squares method

#include <bits/stdc++.h>

#define pb push_back
#define sz(x) (int)x.size()
#define whole(x) x.begin(), x.end()
#define rwhole(x) x.rbegin(), x.rend()

using namespace std;

typedef long long ll;
typedef long double ld;

const int N = (int)3e5 + 25;
const int maxN = (int)1e5 + 2;
const int INF = (int)1e9 + 7;
const double EPS = 1e-7;
const ll linf = (ll)1e17;

double get(int n, double sumX, double sumY, double sumXY, double sumXX) {
    return (n * 1.0 * sumXY - sumX * sumY) / (n * sumXX - (sumX * sumX));
}

double get(int n, double m, double sumY, double sumX) {
    if (sumY - m * sumX < EPS) return 0.0;
    return (sumY - m * sumX) / n;
}

void solve() {
    cout << "Enter the number of points: \n";
    int n; cin >> n;
    cout << "Enter the points:\n";
    vector<double> x(n), y(n);
    for (int i = 0; i < n; ++i) {
        cin >> x[i] >> y[i];
    }

    vector<vector<double>> t(n, vector<double> (4));
    double sumX = 0, sumY = 0, sumXY = 0, sumXX = 0;
    for (int j = 0; j < 4; ++j) {
        if (j == 0) {
            for (int i = 0; i < n; ++i) t[i][j] = x[i], sumX += x[i];
        }
        else if (j == 1) {
            for (int i = 0; i < n; ++i) t[i][j] = y[i], sumY += y[i];
        }
        else if (j == 2) {
            for (int i = 0; i < n; ++i) t[i][j] = x[i] * y[i], sumXY += t[i][j];
        }
        else if (j == 3) {
            for (int i = 0; i < n; ++i) t[i][j] = x[i] * x[i], sumXX += t[i][j];
        }
    }
    cout << "The table is;\n";
    cout << "X  Y  XY  X^2\n";
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < 4; ++j) {
            cout << t[i][j] << " ";
        }
        cout << "\n";
    }

    printf ("SUMS ARE: SUMX: %lf SUMY: %lf SUMXY: %lf SUMX^2: %lf\n", sumX, sumY, sumXY, sumXX);

    double slope = get(n, sumX, sumY, sumXY, sumXX);
    cout << "The slope is: " << slope << "\n";
    cout << "Y - inter: " << get(n, slope, sumY, sumX);
}

int main() {
    int tt = 1;
    while (tt--) solve();
    return 0;
}

File: least_squares.go
package main

import (
	"fmt"
)

func getSlope(n int, sumX float64, sumY float64, sumXY float64, sumXX float64) float64 {
	return (float64(n)*sumXY - sumX*sumY) / (float64(n)*sumXX - (sumX * sumX))
}

func getIntercept(n int, m float64, sumY float64, sumX float64) float64 {
	if sumY-m*sumX < 1e-7 {
		return 0.0
	}
	return (sumY - m*sumX) / float64(n)
}

func main() {
	fmt.Println("Enter the number of points:")
	var n int
	fmt.Scanf("%d", &n)

	fmt.Println("Enter the points:")
	x := make([]float64, n)
	y := make([]float64, n)
	for i := 0; i < n; i++ {
		fmt.Scanf("%f %f", &x[i], &y[i])
	}

	fmt.Println(x, y)

	sumX := 0.0
	sumY := 0.0
	sumXY := 0.0
	sumXX := 0.0
	for i := 0; i < n; i++ {
		sumX += x[i]
		sumY += y[i]
		sumXY += x[i] * y[i]
		sumXX += x[i] * x[i]
	}

	slope := getSlope(n, sumX, sumY, sumXY, sumXX)
	intercept := getIntercept(n, slope, sumY, sumX)

	fmt.Printf("The slope is: %f\n", slope)
	fmt.Printf("The intercept is: %f\n", intercept)
}

//7
//-3 -2 -1 0 1 2 3
// 4.63 2.11 0.67 0.09 0.63 2.15 4.58
